mcp/src/test/java/io/modelcontextprotocol/client/McpAsyncClientResponseHandlerTests.java

```diff
@@ -27,6 +28,83 @@
 
+	private InitializeResult initialization(McpAsyncClient asyncMcpClient, MockMcpTransport transport) {
+		// Create mock server response
+		McpSchema.InitializeResult mockInitResult = new McpSchema.InitializeResult(McpSchema.LATEST_PROTOCOL_VERSION,
+				McpSchema.ServerCapabilities.builder()
+					.tools(true)
+					.resources(true, true) // Enable both resources and resource templates
+					.build(),
+				new McpSchema.Implementation("test-server", "1.0.0"), "Test instructions");
+
+		// Create a Mono that will handle the initialization and response simulation
+		return asyncMcpClient.initialize().doOnSubscribe(subscription -> {
+			// Run in a separate reactive context to avoid blocking the main subscription
+			Mono.fromRunnable(() -> {
+				McpSchema.JSONRPCRequest initRequest = transport.getLastSentMessageAsRequest();
+				assertThat(initRequest.method()).isEqualTo(McpSchema.METHOD_INITIALIZE);
+
+				// Send mock server response
+				McpSchema.JSONRPCResponse initResponse = new McpSchema.JSONRPCResponse(McpSchema.JSONRPC_VERSION,
+						initRequest.id(), mockInitResult, null);
+				transport.simulateIncomingMessage(initResponse);
+			}).subscribeOn(reactor.core.scheduler.Schedulers.boundedElastic()).subscribe();
+		}).block();
+	}
+
+	@Test
+	void testSuccessfulInitialization() {
+		MockMcpTransport transport = new MockMcpTransport();
+		McpAsyncClient asyncMcpClient = McpClient.async(transport).build();
+
+		// Verify client is not initialized initially
+		assertThat(asyncMcpClient.isInitialized()).isFalse();
+
+		// Create mock server response
+		McpSchema.ServerCapabilities mockServerCapabilities = McpSchema.ServerCapabilities.builder()
+			.tools(true)
+			.resources(true, true) // Enable both resources and resource templates
+			.build();
+		McpSchema.Implementation mockServerInfo = new McpSchema.Implementation("test-server", "1.0.0");
+		McpSchema.InitializeResult mockInitResult = new McpSchema.InitializeResult(McpSchema.LATEST_PROTOCOL_VERSION,
+				mockServerCapabilities, mockServerInfo, "Test instructions");
+
+		// Start initialization with reactive handling
+		InitializeResult result = asyncMcpClient.initialize().doOnSubscribe(subscription -> {
+			// Run in a separate reactive context to avoid blocking the main subscription
+			Mono.fromRunnable(() -> {
+				McpSchema.JSONRPCRequest initRequest = transport.getLastSentMessageAsRequest();
+				assertThat(initRequest.method()).isEqualTo(McpSchema.METHOD_INITIALIZE);
+
+				// Send mock server response
+				McpSchema.JSONRPCResponse initResponse = new McpSchema.JSONRPCResponse(McpSchema.JSONRPC_VERSION,
+						initRequest.id(), mockInitResult, null);
+				transport.simulateIncomingMessage(initResponse);
+				// latch.countDown();
+			}).subscribeOn(reactor.core.scheduler.Schedulers.boundedElastic()).subscribe();
+		}).block();
+
+		// Verify initialized notification was sent
+		McpSchema.JSONRPCMessage notificationMessage = transport.getLastSentMessage();
```

Comment:
I am not sure if the MockMcpTransport was designed to have multiple execution of `getLastSentMessage`. It can potentially freeze processing or cause timing issues.